File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/VideoRendererController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.FlutterRtcVideoRenderer
import com.cloudwebrtc.webrtc.TrackRepository
import com.cloudwebrtc.webrtc.proxy.VideoTrackProxy
import com.cloudwebrtc.webrtc.utils.AnyThreadSink
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class VideoRendererController(
    val binaryMessenger: BinaryMessenger,
    private val videoRenderer: FlutterRtcVideoRenderer
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler, IdentifiableController {
    private val channelId: Int = nextChannelId();
    private val methodChannel: MethodChannel = MethodChannel(
        binaryMessenger,
        ChannelNameGenerator.withId("VideoRenderer", channelId)
    )
    private val eventChannel: EventChannel =
        EventChannel(binaryMessenger, ChannelNameGenerator.withId("VideoRendererEvent", channelId))
    private var eventSink: AnyThreadSink? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)

        videoRenderer.setEventListener(object : FlutterRtcVideoRenderer.Companion.EventListener {
            override fun onFirstFrameRendered(id: Long) {
                eventSink?.success(
                    mapOf(
                        "event" to "onFirstFrameRendered",
                        "id" to id
                    )
                )
            }

            override fun onTextureChangeVideoSize(id: Long, height: Int, width: Int) {
                eventSink?.success(
                    mapOf(
                        "event" to "onTextureChangeVideoSize",
                        "id" to id,
                        "width" to width,
                        "height" to height
                    )
                )
            }

            override fun onTextureChangeRotation(id: Long, rotation: Int) {
                eventSink?.success(
                    mapOf(
                        "event" to "onTextureChangeRotation",
                        "id" to id,
                        "rotation" to rotation
                    )
                )
            }

        })
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setSrcObject" -> {
                val trackId: String = call.argument("trackId")!!

                val track = TrackRepository.getTrack(trackId)!!
                val videoTrack = VideoTrackProxy(track)
                videoRenderer.setVideoTrack(videoTrack)

                result.success(null)
            }
        }
    }

    fun asFlutterResult(): Map<String, Any> = mapOf(
        "channelId" to channelId
    )

    override fun onListen(obj: Any?, sink: EventChannel.EventSink?) {
        eventSink = AnyThreadSink(sink)
    }

    override fun onCancel(obj: Any?) {
        eventSink = null
    }
}

File: ios/Classes/controller/VideoRendererController.swift
import Flutter

/// Controller for the `VideoRenderer`.
class VideoRendererController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of `FlutterRtcVideoRenderer`.
  private var renderer: FlutterRtcVideoRenderer

  /// ID of channels created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Event channel for communicating with Flutter side.
  private var eventChannel: FlutterEventChannel

  /// Controller for the `eventChannel` management.
  private var eventController: EventController

  /// Creates a new `VideoRendererController` for the provided `FlutterRtcVideoRenderer`.
  init(messenger: FlutterBinaryMessenger, renderer: FlutterRtcVideoRenderer) {
    let channelName = ChannelNameGenerator.name(name: "VideoRenderer", id: self.channelId)
    let eventChannelName = ChannelNameGenerator.name(
      name: "VideoRendererEvent", id: self.channelId)
    self.messenger = messenger
    self.renderer = renderer
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.eventChannel = FlutterEventChannel(name: eventChannelName, binaryMessenger: messenger)
    self.eventController = EventController()
    self.renderer.subscribe(
      sub: VideoRendererEventController(
        messenger: self.messenger, eventController: self.eventController))
    self.eventChannel.setStreamHandler(eventController)
    self.channel.setMethodCallHandler({ (call, result) in
      self.onMethodCall(call: call, result: result)
    })
  }

  /// Handles all supported Flutter method calls for the `FlutterRtcVideoRenderer`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "setSrcObject":
      let trackId = argsMap!["trackId"] as? String
      var track: MediaStreamTrackProxy?
      if trackId != nil {
        track = MediaStreamTrackStore.tracks[trackId!]
      } else {
        track = nil
      }
      self.renderer.setVideoTrack(newTrack: track)
      result(nil)
    case "dispose":
      self.renderer.dispose()
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller to the Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": channelId,
      "textureId": renderer.getTextureId(),
    ]
  }
}
